src/ProvenanceTransformer.cpp

```diff
@@ -253,7 +254,153 @@ void transformEqrelRelation(AstRelation& rel) {
+                            atom->addArgument(std::make_unique<AstUnnamedVariable>());
+                        }
+                    }
+
+                    // otherwise - apply mapper recursively
+                    node->apply(*this);
+                    return node;
+                }
+            };
+
+            // add unnamed vars to each atom nested in arguments of head
+            clause->getHead()->apply(M());
+
+            // if fact, level number is 0
+            if (clause->isFact()) {
+                clause->getHead()->addArgument(std::make_unique<AstNumberConstant>(0));
+                for (size_t i = 0; i < relation->numberOfHeightParameters(); i++)
+                    clause->getHead()->addArgument(std::make_unique<AstNumberConstant>(0));
+            } else {
+                std::vector<AstArgument*> bodyLevels;
+
+                for (size_t i = 0; i < clause->getBodyLiterals().size(); i++) {
+                    auto lit = clause->getBodyLiterals()[i];
+
+                    // add unnamed vars to each atom nested in arguments of lit
+                    lit->apply(M());
+
+                    // add provenance columns to lit; first rule num, then level nums
+                    if (auto atom = dynamic_cast<AstAtom*>(lit)) {
+                        // rule num
+                        atom->addArgument(std::make_unique<AstUnnamedVariable>());
+                        atom->addArgument(
+                                std::make_unique<AstVariable>("@level_number_" + std::to_string(i)));
+                        // level nums
+                        for (size_t j = 0;
+                                j < program->getRelation(atom->getName())->numberOfHeightParameters() - 1;
+                                j++) {
+                            atom->addArgument(std::make_unique<AstUnnamedVariable>());
+                        }
+                        bodyLevels.push_back(new AstVariable("@level_number_" + std::to_string(i)));
+                    }
+                }
+
+                // add provenance columns to head lit
+                // rule number
+                clause->getHead()->addArgument(std::make_unique<AstNumberConstant>(clause->getClauseNum()));
+                // max level
+                clause->getHead()->addArgument(std::unique_ptr<AstArgument>(getNextLevelNumber(bodyLevels)));
+                // level numbers
+                for (size_t j = 0; j < clause->getAtoms().size(); j++) {
+                    clause->getHead()->addArgument(
+                            std::make_unique<AstVariable>("@level_number_" + std::to_string(j)));
+                }
+                for (size_t j = clause->getAtoms().size(); j < relation->numberOfHeightParameters() - 1;
+                        j++) {
+                    clause->getHead()->addArgument(std::make_unique<AstNumberConstant>(
+                            -1));  // which value to use for encoding empty height parameter?
```

Comment:
Avoid end of line comments, and especially avoid long end of line comments. Move to the line before.

Also, is the comment still needed?

Reply:
Comment removed in clean up commit.